File: VB/E497/CustomNavigation.vb
Imports Microsoft.VisualBasic
Imports DevExpress.Web.ASPxScheduler
Imports DevExpress.Web.ASPxScheduler.Internal
Imports DevExpress.XtraScheduler
Imports DevExpress.XtraScheduler.Commands
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Data
Imports System.Diagnostics
Imports System.Threading

Public Class MonthViewHelper
	Public Shared Function CalculateWeekCount(ByVal scheduler As ASPxScheduler, ByVal monthStart As DateTime) As Integer
		Dim localDoW As DayOfWeek = If(scheduler.OptionsView.FirstDayOfWeek = DevExpress.XtraScheduler.FirstDayOfWeek.System, Thread.CurrentThread.CurrentCulture.DateTimeFormat.FirstDayOfWeek, CType(scheduler.OptionsView.FirstDayOfWeek, DayOfWeek))

		Dim weekStart As DateTime = FirstDayOfWeek(monthStart, localDoW)
		Dim monthEnd As DateTime = monthStart.AddMonths(1)
		Dim weekCount As Integer = 0
		Do While weekStart < monthEnd
			weekCount += 1
			weekStart = weekStart.AddDays(7)
		Loop
		Return weekCount
	End Function

	Public Shared Function FirstDayOfWeek(ByVal [date] As DateTime, ByVal startOfWeek As DayOfWeek) As DateTime
		Dim diff As Integer = (7 + ([date].DayOfWeek - startOfWeek)) Mod 7
		Return [date].AddDays(-1 * diff).Date
	End Function
End Class

Public Class CustomOffsetVisibleIntervalsCallbackCommand
	Inherits OffsetVisibleIntervalsCallbackCommand
	Public Sub New(ByVal scheduler As ASPxScheduler)
		MyBase.New(scheduler)
	End Sub

	Protected Overrides Sub ExecuteCore()
		'base.ExecuteCore();
		Dim currentMonth As DateTime = FirstDate.AddDays(15)
		currentMonth = New DateTime(currentMonth.Year, currentMonth.Month, 1)
		Control.Start = currentMonth
		Control.MonthView.WeekCount = MonthViewHelper.CalculateWeekCount(Control, currentMonth)
	End Sub
End Class
Public Class CustomNavigateForwardCallbackCommand
	Inherits NavigateForwardCallbackCommand
	Public Sub New(ByVal scheduler As ASPxScheduler)
		MyBase.New(scheduler)
	End Sub

	Protected Overrides Sub ExecuteCore()
		'base.ExecuteCore();
		Dim currentMonth As DateTime = Control.Start.AddDays(15)
		currentMonth = New DateTime(currentMonth.Year, currentMonth.Month, 1)
		Dim newStartDate As DateTime = currentMonth.AddMonths(1)
		Control.Start = newStartDate
		Control.MonthView.WeekCount = MonthViewHelper.CalculateWeekCount(Control, newStartDate)
	End Sub
End Class

Public Class CustomNavigateBackwardCallbackCommand
	Inherits NavigateBackwardCallbackCommand
	Public Sub New(ByVal scheduler As ASPxScheduler)
		MyBase.New(scheduler)
	End Sub

	Protected Overrides Sub ExecuteCore()
		'base.ExecuteCore();
		Dim currentMonth As DateTime = Control.Start.AddDays(15)
		currentMonth = New DateTime(currentMonth.Year, currentMonth.Month, 1)
		Dim newStartDate As DateTime = currentMonth.AddMonths(-1)
		Control.Start = newStartDate
		Control.MonthView.WeekCount = MonthViewHelper.CalculateWeekCount(Control, newStartDate)
	End Sub
End Class

Public Class CustomGotoTodayCallbackCommand
	Inherits GotoTodayCallbackCommand
	Public Sub New(ByVal scheduler As ASPxScheduler)
		MyBase.New(scheduler)
	End Sub

	Protected Overrides Sub ExecuteCore()
		'base.ExecuteCore();
		Dim newStartDate As New DateTime(DateTime.Now.Year, DateTime.Now.Month, 1)
		Control.Start = newStartDate
		Control.MonthView.WeekCount = MonthViewHelper.CalculateWeekCount(Control, newStartDate)
	End Sub
End Class

Public Class CustomGotoDateDialogCommand
	Inherits DevExpress.Web.ASPxScheduler.Internal.GotoDateFormCallbackCommand
	Public Sub New(ByVal scheduler As ASPxScheduler)
		MyBase.New(scheduler)
	End Sub

	Protected Overrides Sub ExecuteCore()
		Dim newStartDate As New DateTime(NewDate.Date.Year, NewDate.Date.Month, 1)
		Control.Start = newStartDate
		Control.MonthView.WeekCount = MonthViewHelper.CalculateWeekCount(Control, newStartDate)
	End Sub
End Class

Public Class CustomGotoDateCallbackCommand
	Inherits GotoDateCallbackCommand
	Public Sub New(ByVal scheduler As ASPxScheduler)
		MyBase.New(scheduler)
	End Sub

	Protected Overrides Sub ExecuteCore()
		'base.ExecuteCore();
		Dim newStartDate As New DateTime(NewDate.Year, NewDate.Month, 1)
		Control.Start = newStartDate
		Control.MonthView.WeekCount = MonthViewHelper.CalculateWeekCount(Control, newStartDate)
	End Sub
End Class

Public Class CustomSwitchViewCallbackCommand
	Inherits SwitchViewCallbackCommand
	Public Sub New(ByVal scheduler As ASPxScheduler)
		MyBase.New(scheduler)
	End Sub

	Protected Overrides Sub ExecuteCore()
		MyBase.ExecuteCore()
		If Control.ActiveViewType = SchedulerViewType.Month Then
			Dim newStartDate As New DateTime(Control.Start.Year, Control.Start.Month, 1)
			Control.Start = newStartDate
			Control.MonthView.WeekCount = MonthViewHelper.CalculateWeekCount(Control, newStartDate)
		End If
	End Sub
End Class

File: CS/E497/CustomNavigation.cs

using DevExpress.Web.ASPxScheduler;
using DevExpress.Web.ASPxScheduler.Internal;
using DevExpress.XtraScheduler;
using DevExpress.XtraScheduler.Commands;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Threading;

public class MonthViewHelper {
    public static int CalculateWeekCount(ASPxScheduler scheduler, DateTime monthStart) {
        DayOfWeek localDoW = scheduler.OptionsView.FirstDayOfWeek == DevExpress.XtraScheduler.FirstDayOfWeek.System ? Thread.CurrentThread.CurrentCulture.DateTimeFormat.FirstDayOfWeek :
            (DayOfWeek)scheduler.OptionsView.FirstDayOfWeek;

        DateTime weekStart = FirstDayOfWeek(monthStart, localDoW);
        DateTime monthEnd = monthStart.AddMonths(1);
        int weekCount = 0;
        while(weekStart < monthEnd) {
            weekCount++;
            weekStart = weekStart.AddDays(7);
        }
        return weekCount;
    }

    public static DateTime FirstDayOfWeek(DateTime date, DayOfWeek startOfWeek) {
        int diff = (7 + (date.DayOfWeek - startOfWeek)) % 7;
        return date.AddDays(-1 * diff).Date;
    }
}

public class CustomOffsetVisibleIntervalsCallbackCommand : OffsetVisibleIntervalsCallbackCommand {
    public CustomOffsetVisibleIntervalsCallbackCommand(ASPxScheduler scheduler) : base(scheduler) { }

    protected override void ExecuteCore() {
        //base.ExecuteCore();
        DateTime currentMonth = FirstDate.AddDays(15);
        currentMonth = new DateTime(currentMonth.Year, currentMonth.Month, 1);
        Control.Start = currentMonth;
        Control.MonthView.WeekCount = MonthViewHelper.CalculateWeekCount(Control, currentMonth);
    }
}
public class CustomNavigateForwardCallbackCommand : NavigateForwardCallbackCommand {
    public CustomNavigateForwardCallbackCommand(ASPxScheduler scheduler) : base(scheduler) { }

    protected override void ExecuteCore() {
        //base.ExecuteCore();
        DateTime currentMonth = Control.Start.AddDays(15);
        currentMonth = new DateTime(currentMonth.Year, currentMonth.Month, 1);
        DateTime newStartDate = currentMonth.AddMonths(1);
        Control.Start = newStartDate;
        Control.MonthView.WeekCount = MonthViewHelper.CalculateWeekCount(Control, newStartDate);
    }
}

public class CustomNavigateBackwardCallbackCommand : NavigateBackwardCallbackCommand {
    public CustomNavigateBackwardCallbackCommand(ASPxScheduler scheduler) : base(scheduler) { }

    protected override void ExecuteCore() {
        //base.ExecuteCore();
        DateTime currentMonth = Control.Start.AddDays(15);
        currentMonth = new DateTime(currentMonth.Year, currentMonth.Month, 1);
        DateTime newStartDate = currentMonth.AddMonths(-1);
        Control.Start = newStartDate;
        Control.MonthView.WeekCount = MonthViewHelper.CalculateWeekCount(Control, newStartDate);
    }
}

public class CustomGotoTodayCallbackCommand : GotoTodayCallbackCommand {
    public CustomGotoTodayCallbackCommand(ASPxScheduler scheduler) : base(scheduler) { }

    protected override void ExecuteCore() {
        //base.ExecuteCore();
        DateTime newStartDate = new DateTime(DateTime.Now.Year, DateTime.Now.Month, 1);
        Control.Start = newStartDate;
        Control.MonthView.WeekCount = MonthViewHelper.CalculateWeekCount(Control, newStartDate);
    }
}

public class CustomGotoDateDialogCommand : DevExpress.Web.ASPxScheduler.Internal.GotoDateFormCallbackCommand {
    public CustomGotoDateDialogCommand(ASPxScheduler scheduler) : base(scheduler) { }

    protected override void ExecuteCore() {
        DateTime newStartDate = new DateTime(NewDate.Date.Year, NewDate.Date.Month, 1);
        Control.Start = newStartDate;
        Control.MonthView.WeekCount = MonthViewHelper.CalculateWeekCount(Control, newStartDate);
    }
}

public class CustomGotoDateCallbackCommand : GotoDateCallbackCommand {
    public CustomGotoDateCallbackCommand(ASPxScheduler scheduler) : base(scheduler) { }

    protected override void ExecuteCore() {
        //base.ExecuteCore();
        DateTime newStartDate = new DateTime(NewDate.Year, NewDate.Month, 1);
        Control.Start = newStartDate;
        Control.MonthView.WeekCount = MonthViewHelper.CalculateWeekCount(Control, newStartDate);
    }
}

public class CustomSwitchViewCallbackCommand : SwitchViewCallbackCommand {
    public CustomSwitchViewCallbackCommand(ASPxScheduler scheduler) : base(scheduler) { }

    protected override void ExecuteCore() {
        base.ExecuteCore();
        if(Control.ActiveViewType == SchedulerViewType.Month) {
            DateTime newStartDate = new DateTime(Control.Start.Year, Control.Start.Month, 1);
            Control.Start = newStartDate;
            Control.MonthView.WeekCount = MonthViewHelper.CalculateWeekCount(Control, newStartDate);
        }
    }
}